sdks/python/apache_beam/io/range_trackers.py

```diff
@@ -324,3 +324,123 @@ def set_current_position(self, position):
 
   def fraction_consumed(self):
     return self._range_tracker.fraction_consumed()
+
+
+class LexicographicKeyRangeTracker(iobase.RangeTracker):
+  """
+  A range tracker that tracks progress through a lexicographically
+  ordered keyspace of strings.
+  """
+
+  def __init__(self, start_key=None, stop_key=None):
+    self._start_key = start_key
+    self._stop_key = stop_key
+    self._lock = threading.Lock()
+    self._last_claim = None
+
+  def start_position(self):
+    return self._start_key
+
+  def stop_position(self):
+    return self._stop_key
+
+  def try_claim(self, position):  # pylint: disable=unused-argument
+    with self._lock:
+      if self.stop_key is None or position < self.stop_key:
+        self._last_claim = position
+        return True
+
+  def position_at_fraction(self, fraction):
+    return self._fraction_to_key(fraction, self._start_key, self._end_key)
+
+  def try_split(self, position):
+    with self._lock:
+      if self._last_claim < position:
+        fraction = self._key_to_fraction(
+            position, self._start_key, self._stop_key)
+        self._stop_key = position
+        return position, fraction
+      else:
+        return None
+
+  def fraction_consumed(self):
+    if self._last_claim is None:
+      return 0
+    else:
+      return self._key_to_fraction(
+          self._last_claim, self._start_key, self._stop_key)
+
+  @classmethod
+  def _fraction_to_key(cls, fraction, start=None, end=None):
+    """
+    Linearly interpolates a key that is lexicographically
+    fraction of the way between start and end.
+    """
+    assert 0 <= fraction <= 1, fraction
+    if start is None:
+      start = ''
+    if fraction == 1:
+      return end
+    elif fraction == 0:
+      return start
+    else:
+      if not end:
```

Comment:
Actually, I just implemented this based on an idea I had on the plane coming back from MTV. But looking at the implementation in Java, it's quite similar, though Python's bigints are more convenient to deal with than Java's and I just use hex to convert (as this isn't performance critical).